File: tetris_cpp/src/gamepool/block/block.cpp
#include "block.h"
#include "../PoolConstant.h"
#include "../../window/RunParameter.h"
#include <QDebug>
#include <QPainter>
#include <QPixmap>
#include <cstdlib>

Block::Block(int shape)
{/*{{{*/
	if (shape < 0) return;
	this->shape = shape;
	switch (shape) {
		case 0:
			{
				this->x = 2;
				this->y = 1;
				this->img = "T";
				this->initT();
				break;
			}
		case 1:
			{
				this->x = 2.5;
				this->y = 1.5;
				this->img = "L";
				this->initL();
				break;
			}
		case 2:
			{
				this->x = 1.5;
				this->y = 1.5;
				this->img = "J";
				this->initJ();
				break;
			}
		case 3:
			{
				this->x = 2;
				this->y = 2;
				this->img = "Z";
				this->initZ();
				break;
			}
		case 4:
			{
				this->x = 2;
				this->y = 2;
				this->img = "S";
				this->initS();
				break;
			}
		case 5:
			{
				this->x = 1.5;
				this->y = 1.5;
				this->img = "I";
				this->initI();
				break;
			}
		case 6:
			{
				this->x = 1.5;
				this->y = 1;
				this->img = "O";
				this->initO();
				break;
			}
	}
}/*}}}*/

void Block::draw(QPainter &p)
{
	if (gameover) p.setOpacity(0.5);
	QPixmap cell;
	cell.load(IMG_PATH + "" + img + ".png");
	for (auto &point : cells) {
		p.drawPixmap((*point) * CELL_WIDTH, (*(point + 1)) * CELL_WIDTH, CELL_WIDTH, CELL_WIDTH, cell);
	}
}

void Block::turnD()
{
	++status;
	initShape();
	if (!checkPool()) {
		--status;
		initShape();
	}
}

void Block::turnS()
{
	--status;
	initShape();
	if (!checkPool()) {
		++status;
		initShape();
	}
}

bool Block::moveLeft()
{
	--x;
	initShape();
	if (!checkPool()) {
		++x;
		initShape();
		return false;
	}
	return true;
}
void Block::moveLeft3()
{
	int t = 0;
	while (t++ < 3 && moveLeft());
}
void Block::moveLeftEnd()
{
	while (moveLeft());
}

bool Block::moveRight()
{
	++x;
	initShape();
	if (!checkPool()) {
		--x;
		initShape();
		return false;
	}
	return true;
}
void Block::moveRight3()
{
	int t = 0;
	while (t++ < 3 && moveRight());
}
void Block::moveRightEnd()
{
	while (moveRight());
}

void Block::moveTop()
{
	--y;
	initShape();
}

bool Block::moveDown()
{
	++y;
	initShape();
	if (!checkPool()) {
		if (y) --y;
		initShape();
		return false;
	}
	return true;
}

void Block::moveBottom()
{
	while (moveDown());
}

void Block::putBlock()
{
	for (auto &point : cells)
	{
		poolMap[int(*(point + 1))][int(*point)] = img;
	}
}

void Block::clearBlock()
{
	for (auto &point : cells)
	{
		poolMap[int(*(point + 1))][int(*point)] = "";
	}
}

void Block::setShow()
{
	this->x = 5;
	this->y = 1;
	this->status = 0;
	initShape();
}

bool Block::checkPool()
{
	for (auto &point : cells)
		if ((*point) < 0 || (*point) >= POOL_W) return false;
		else if (*(point + 1) < 0 || *(point + 1) >= POOL_H) return false;
	for (int i = 0; i < POOL_H; ++i)
		for (int j = 0; j < POOL_W; ++j)
			if (!poolMap[i][j].isEmpty()) {
				for (auto &point : cells)
					if ((*point) == j && *(point + 1) == i) {
						return false;
					}
			}
	return true;
}

void Block::clearRepeatCell()
{
	for (auto &dpoint : cells)
		poolMap[(int)(*(dpoint + 1))][(int)(*dpoint)] = "";
}

Block Block::clone()
{
	return Block(shape);
}

void Block::initShape()
{
	int maxStatus = 3;
	if (shape > 2) maxStatus = 1;

	if (shape == 6)
		status = 0;
	else {
		if (status > maxStatus) status = 0;
		else if (status < 0) status = maxStatus;
	}

	switch (shape) {
		case 0: initT(); break;
		case 1: initL(); break;
		case 2: initJ(); break;
		case 3: initZ(); break;
		case 4: initS(); break;
		case 5: initI(); break;
		case 6: initO(); break;
	}
}

void Block::initT()
{
	cells[0][0] = x;
	cells[0][1] = y;
	if (status == 0) {
		cells[1][0] = x - 1;
		cells[1][1] = y;
		cells[2][0] = x + 1;
		cells[2][1] = y;
		cells[3][0] = x;
		cells[3][1] = y + 1;
	} else if (status == 1) {
		cells[1][0] = x;
		cells[1][1] = y - 1;
		cells[2][0] = x - 1;
		cells[2][1] = y;
		cells[3][0] = x;
		cells[3][1] = y + 1;
	} else if (status == 2) {
		cells[1][0] = x;
		cells[1][1] = y - 1;
		cells[2][0] = x - 1;
		cells[2][1] = y;
		cells[3][0] = x + 1;
		cells[3][1] = y;
	} else if (status == 3) {
		cells[1][0] = x;
		cells[1][1] = y - 1;
		cells[2][0] = x + 1;
		cells[2][1] = y;
		cells[3][0] = x;
		cells[3][1] = y + 1;
	}
}

void Block::initL()
{
	cells[0][0] = x;
	cells[0][1] = y;
	if (status == 0) {
		cells[1][0] = x - 1;
		cells[1][1] = y - 1;
		cells[2][0] = x;
		cells[2][1] = y - 1;
		cells[3][0] = x;
		cells[3][1] = y + 1;
	} else if (status == 1) {
		cells[1][0] = x + 1;
		cells[1][1] = y - 1;
		cells[2][0] = x - 1;
		cells[2][1] = y;
		cells[3][0] = x + 1;
		cells[3][1] = y;
	} else if (status == 2) {
		cells[1][0] = x;
		cells[1][1] = y - 1;
		cells[2][0] = x;
		cells[2][1] = y + 1;
		cells[3][0] = x + 1;
		cells[3][1] = y + 1;
	} else if (status == 3) {
		cells[1][0] = x - 1;
		cells[1][1] = y;
		cells[2][0] = x + 1;
		cells[2][1] = y;
		cells[3][0] = x - 1;
		cells[3][1] = y + 1;
	}
}

void Block::initJ()
{
	cells[0][0] = x;
	cells[0][1] = y;
	if (status == 0) {
		cells[1][0] = x + 1;
		cells[1][1] = y - 1;
		cells[2][0] = x;
		cells[2][1] = y - 1;
		cells[3][0] = x;
		cells[3][1] = y + 1;
	} else if (status == 1) {
		cells[1][0] = x + 1;
		cells[1][1] = y + 1;
		cells[2][0] = x - 1;
		cells[2][1] = y;
		cells[3][0] = x + 1;
		cells[3][1] = y;
	} else if (status == 2) {
		cells[1][0] = x;
		cells[1][1] = y - 1;
		cells[2][0] = x;
		cells[2][1] = y + 1;
		cells[3][0] = x - 1;
		cells[3][1] = y + 1;
	} else if (status == 3) {
		cells[1][0] = x - 1;
		cells[1][1] = y;
		cells[2][0] = x + 1;
		cells[2][1] = y;
		cells[3][0] = x - 1;
		cells[3][1] = y - 1;
	}
}

void Block::initZ()
{
	cells[0][0] = x;
	cells[0][1] = y;
	if (status == 0) {
		cells[1][0] = x - 1;
		cells[1][1] = y - 1;
		cells[2][0] = x;
		cells[2][1] = y - 1;
		cells[3][0] = x + 1;
		cells[3][1] = y;
	} else if (status == 1) {
		cells[1][0] = x + 1;
		cells[1][1] = y - 1;
		cells[2][0] = x + 1;
		cells[2][1] = y;
		cells[3][0] = x;
		cells[3][1] = y + 1;
	}
}

void Block::initS()
{
	cells[0][0] = x;
	cells[0][1] = y;
	if (status == 0) {
		cells[1][0] = x + 1;
		cells[1][1] = y - 1;
		cells[2][0] = x;
		cells[2][1] = y - 1;
		cells[3][0] = x - 1;
		cells[3][1] = y;
	} else if (status == 1) {
		cells[1][0] = x;
		cells[1][1] = y - 1;
		cells[2][0] = x + 1;
		cells[2][1] = y;
		cells[3][0] = x + 1;
		cells[3][1] = y + 1;
	}
}

void Block::initI()
{
	cells[0][0] = x;
	cells[0][1] = y;
	if (status == 0) {
		cells[1][0] = x + 1;
		cells[1][1] = y;
		cells[2][0] = x + 2;
		cells[2][1] = y;
		cells[3][0] = x - 1;
		cells[3][1] = y;
	} else if (status == 1) {
		cells[0][0] = x + 1;
		cells[1][0] = x + 1;
		cells[1][1] = y - 1;
		cells[2][0] = x + 1;
		cells[2][1] = y + 1;
		cells[3][0] = x + 1;
		cells[3][1] = y + 2;
	}
}

void Block::initO()
{
	cells[0][0] = x;
	cells[0][1] = y;
	cells[1][0] = x + 1;
	cells[1][1] = y;
	cells[2][0] = x + 1;
	cells[2][1] = y + 1;
	cells[3][0] = x;
	cells[3][1] = y + 1;
}

File: tetris_python/block/block.py
# -*- coding: utf-8 -*-
import random
from PyQt5.QtGui import QPixmap

CELL_WIDTH = 22

class Block():

    def __init__(self, shape=-1):
        self.x = float(0)
        self.y = float(0)
        self.cells = []
        self.img = ''
        if shape == -1:
            shape = random.randint(0, 6)
        self.shape = shape
        self.status = 0

        self.pool_blocks = []

        self.__new_by_shape()

    def __new_by_shape(self):
        if self.shape == 0:
            self.x = 2
            self.y = 1
            self.img = './T.png'
            self.__init_t()
        elif self.shape == 1:
            self.x = 2.5
            self.y = 1.5
            self.img = './L.png'
            self.__init_l()
        elif self.shape == 2:
            self.x = 1.5
            self.y = 1.5
            self.img = './J.png'
            self.__init_j()
        elif self.shape == 3:
            self.x = 2
            self.y = 2
            self.img = './Z.png'
            self.__init_z()
        elif self.shape == 4:
            self.x = 2
            self.y = 2
            self.img = './S.png'
            self.__init_s()
        elif self.shape == 5:
            self.x = 1.5
            self.y = 1.5
            self.img = './I.png'
            self.__init_i()
        elif self.shape == 6:
            self.x = 1.5
            self.y = 1
            self.img = './O.png'
            self.__init_o()


    def __init_t(self):
        x = self.x
        y = self.y
        cells = []
        cells.append([x, y])
        if self.status == 0:
            cells.append([x - 1, y])
            cells.append([x + 1, y])
            cells.append([x, y + 1])
        elif self.status == 1:
            cells.append([x, y - 1])
            cells.append([x - 1, y])
            cells.append([x, y + 1])
        elif self.status == 2:
            cells.append([x, y - 1])
            cells.append([x - 1, y])
            cells.append([x + 1, y])
        elif self.status == 3:
            cells.append([x, y - 1])
            cells.append([x + 1, y])
            cells.append([x, y + 1])
        self.cells = cells


    def __init_l(self):
        x = self.x
        y = self.y
        cells = []
        cells.append([x, y])
        if self.status == 0:
            cells.append([x - 1, y - 1])
            cells.append([x, y - 1])
            cells.append([x, y + 1])
        elif self.status == 1:
            cells.append([x + 1, y - 1])
            cells.append([x - 1, y])
            cells.append([x + 1, y])
        elif self.status == 2:
            cells.append([x, y - 1])
            cells.append([x, y + 1])
            cells.append([x + 1, y + 1])
        elif self.status == 3:
            cells.append([x - 1, y])
            cells.append([x + 1, y])
            cells.append([x - 1, y + 1])
        self.cells = cells

    def __init_j(self):
        x = self.x
        y = self.y
        cells = []
        cells.append([x, y])
        if self.status == 0:
            cells.append([x + 1, y - 1])
            cells.append([x, y - 1])
            cells.append([x, y + 1])
        elif self.status == 1:
            cells.append([x + 1, y + 1])
            cells.append([x - 1, y])
            cells.append([x + 1, y])
        elif self.status == 2:
            cells.append([x, y - 1])
            cells.append([x, y + 1])
            cells.append([x - 1, y + 1])
        elif self.status == 3:
            cells.append([x - 1, y])
            cells.append([x + 1, y])
            cells.append([x - 1, y - 1])
        self.cells = cells

    def __init_z(self):
        x = self.x
        y = self.y
        cells = []
        cells.append([x, y])
        if self.status == 0:
            cells.append([x - 1, y - 1])
            cells.append([x, y - 1])
            cells.append([x + 1, y])
        elif self.status == 1:
            cells.append([x + 1, y - 1])
            cells.append([x + 1, y])
            cells.append([x, y + 1])
        self.cells = cells

    def __init_s(self):
        x = self.x
        y = self.y
        cells = []
        cells.append([x, y])
        if self.status == 0:
            cells.append([x + 1, y - 1])
            cells.append([x, y - 1])
            cells.append([x - 1, y])
        elif self.status == 1:
            cells.append([x, y - 1])
            cells.append([x + 1, y])
            cells.append([x + 1, y + 1])
        self.cells = cells

    def __init_i(self):
        x = self.x
        y = self.y
        cells = []
        cells.append([x, y])
        if self.status == 0:
            cells.append([x + 1, y])
            cells.append([x + 2, y])
            cells.append([x - 1, y])
        elif self.status == 1:
            cells.append([x + 1, y - 1])
            cells.append([x + 1, y + 1])
            cells.append([x + 1, y + 2])
            cells[0][0] = x + 1
        self.cells = cells

    def __init_o(self):
        x = self.x
        y = self.y
        cells = []
        cells.append([x, y])
        cells.append([x + 1, y])
        cells.append([x + 1, y + 1])
        cells.append([x, y + 1])
        self.cells = cells


    def move_left(self):
        self.x -= 1
        self.__reset_block()
        if not self.__check_border() or not self.__check_pool_blocks():
            self.move_right()
            return False
        return True
    def move_left_3(self):
        times = 0
        while times < 3 and self.move_left():
            times += 1
    def move_left_end(self):
        while self.move_left():
            pass


    def move_right(self):
        self.x += 1
        self.__reset_block()
        if not self.__check_border() or not self.__check_pool_blocks():
            self.move_left()
            return False
        return True
    def move_right_3(self):
        times = 0
        while times < 3 and self.move_right():
            times += 1
    def move_right_end(self):
        while self.move_right():
            pass


    def move_down(self):
        self.y += 1
        self.__reset_block()
        if not self.__check_border() or not self.__check_pool_blocks():
            self.move_top()
            return False
        return True


    def move_bottom(self):
        while self.move_down():
            pass


    def move_top(self):
        self.y -= 1
        self.__reset_block()
        if not self.__check_border():
            self.move_down()


    def turn_d(self):
        self.status += 1
        self.__reset_block()
        if not self.__check_border() or not self.__check_pool_blocks():
            self.turn_s()


    def turn_s(self):
        self.status -= 1
        self.__reset_block()
        if not self.__check_border() or not self.__check_pool_blocks():
            self.turn_d()


    def show_in_pool(self, pool_blocks):
        self.x = 5
        self.y = 1
        self.status = 0
        self.__reset_block()
        self.pool_blocks = pool_blocks


    def __check_pool_blocks(self):
        for cell in self.cells:
            if self.pool_blocks[cell[1]][cell[0]]:
                return False
        return True


    def __check_border(self):
        for cell in self.cells:
            if cell[0] < 0 or cell[0] > 11 or cell[1] < 0 or cell[1] > 20:
                return False

        return True


    def __reset_block(self):
        max_status = 3
        if self.shape > 2:
            max_status = 1

        if self.shape == 6:
            self.status = 0
        else:
            if self.status > max_status:
                self.status = 0
            elif self.status < 0:
                self.status = max_status

        if self.shape == 0:
            self.__init_t()
        elif self.shape == 1:
            self.__init_l()
        elif self.shape == 2:
            self.__init_j()
        elif self.shape == 3:
            self.__init_z()
        elif self.shape == 4:
            self.__init_s()
        elif self.shape == 5:
            self.__init_i()
        elif self.shape == 6:
            self.__init_o()


    def draw_block(self, painter):
        pixmap = QPixmap('./block/' + self.img)
        for cell in self.cells:
            painter.drawPixmap(cell[0] * CELL_WIDTH, cell[1] * CELL_WIDTH,\
                    CELL_WIDTH, CELL_WIDTH, pixmap)
